File: 2024/day_2/solution.py
def read():
    with open("input.txt") as f:
        return f.read()

def check_report(report):
    last_diff = report[1] - report[0]
    if abs(last_diff) > 3 or abs(last_diff) < 1:
        return 0
    for i in range(len(report[2:])):
        diff = report[i+2] - report[i+1]
        if diff*last_diff < 0 or abs(diff) > 3 or abs(diff) < 1:
            return 0
        last_diff = diff
    return 1

if __name__ == "__main__":
    text = read()
    first_list = []
    second_list = []

    reports = []
    safe_1 = 0
    safe_2 = 0

    for line in text.splitlines():
        line_report = [int(i) for i in line.split()]
        safe_report = check_report(line_report)
        if safe_report:
            print(line_report)
        safe_1 += safe_report
        if safe_report == 0:
            for level in range(len(line_report)):
                modified_line_report = line_report.copy()
                del modified_line_report[level]
                safe_report_2 = check_report(modified_line_report)
                if safe_report_2 > 0:
                    safe_2 += 1
                    break
    print(safe_1)
    print(safe_2 + safe_1)

File: 2024/day_2/solution.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"strconv"
	"strings"
)

func Solve(fname string) {
	reports := ReadInput(fname)
	solution1 := Part1(reports)
	solution2 := Part2(reports)

	fmt.Printf("Part 1: %d\nPart 2: %d\n", solution1, solution2)
}

func check_report(report []int) int {
	diff := report[1] - report[0]
	if diff > 3 || diff < -3 || diff == 0 {
		return 0
	}
	for level := 2; level < len(report); level++ {
		current_diff := report[level] - report[level-1]
		if current_diff > 3 || current_diff < -3 || current_diff == 0 || diff*current_diff < 0 {
			return 0
		}
		diff = current_diff
	}
	return 1
}

func Part1(reports [][]int) int {
	safe_report_counter := 0
	for _, report := range reports {
		safe_report := check_report(report)
		safe_report_counter += safe_report
	}
	return safe_report_counter
}

func Part2(reports [][]int) int {
	safe_report_counter := 0
	for _, report := range reports {
		safe_report := check_report(report)
		if safe_report == 0 {
			fmt.Println("---")
			for level := range report {
				new_report := append([]int{}, report...)
				new_report = append(new_report[:level], new_report[level+1:]...)
				fmt.Println(report, new_report)
				safe_report = check_report(new_report)
				if safe_report > 0 {
					safe_report = 1
					break
				}
			}
		}
		safe_report_counter += safe_report
	}
	return safe_report_counter
}

func ReadInput(fname string) [][]int {
	file, err := os.Open(fname)
	if err != nil {
		log.Fatal(err)
	}
	defer file.Close()

	var reports [][]int
	var report []int

	scan := bufio.NewScanner(file)
	for scan.Scan() {
		string_line := strings.Fields(scan.Text())
		report = make([]int, 0)
		for _, part := range string_line {
			num, err := strconv.Atoi(strings.TrimSpace(part))
			if err != nil {
				fmt.Printf("Error converting %s to int: %v\n", part, err)
				continue
			}
			report = append(report, num)
		}
		reports = append(reports, report)
	}

	return reports
}

func main() {
	path, _ := os.Getwd()
	filepath := path + "/input.txt"
	Solve(filepath)
}
